// ngx-editor/esm2022/lib/pipes/sanitize/sanitize-html.pipe.mjs
import { Pipe } from '@angular/core';
import { isTrustedHtml } from '../../trustedTypesUtil';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
export class SanitizeHtmlPipe {
    sanitizer;
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
    }
    transform(value) {
        if (isTrustedHtml(value)) {
            return value;
        }
        return this.sanitizer.bypassSecurityTrustHtml(value);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.2", ngImport: i0, type: SanitizeHtmlPipe, deps: [{ token: i1.DomSanitizer }], target: i0.ɵɵFactoryTarget.Pipe });
    static ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.2.2", ngImport: i0, type: SanitizeHtmlPipe, name: "sanitizeHtml" });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.2", ngImport: i0, type: SanitizeHtmlPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'sanitizeHtml',
                }]
        }], ctorParameters: () => [{ type: i1.DomSanitizer }] });
